zoltar-metrics/src/main/java/com/spotify/zoltar/metrics/semantic/SemanticPredictorMetrics.java

```diff
@@ -40,40 +45,73 @@
 @AutoValue
 public abstract class SemanticPredictorMetrics implements PredictorMetrics {
 
-  abstract Timer predictDurationTimer();
-
-  abstract Meter predictRateCounter();
-
-  abstract Timer extractDurationTimer();
-
-  abstract Meter extractRateCounter();
+  abstract LoadingCache<Id, Meters> metersCache();
 
   /** Creates a new @{link SemanticPredictorMetrics}. */
-  public static SemanticPredictorMetrics create(final SemanticMetricRegistry registry) {
-    final MetricId predictDurationId = MetricId.build().tagged("what", PREDICT_DURATION.tag());
-    final MetricId predictRateId = MetricId.build().tagged("what", PREDICT_RATE.tag());
-    final MetricId extractDurationId = MetricId.build()
-        .tagged("what", FEATURE_EXTRACT_DURATION.tag());
-    final MetricId extractRateId = MetricId.build().tagged("what", FEATURE_EXTRACT_RATE.tag());
-
-    final Timer predictDurationTimer = registry.timer(predictDurationId);
-    final Meter predictRateCounter = registry.meter(predictRateId);
-    final Timer extractDurationTimer = registry.timer(extractDurationId);
-    final Meter extractRateCounter = registry.meter(extractRateId);
-
-    return new AutoValue_SemanticPredictorMetrics(predictDurationTimer,
-                                                  predictRateCounter,
-                                                  extractDurationTimer,
-                                                  extractRateCounter);
+  public static SemanticPredictorMetrics create(final SemanticMetricRegistry registry,
+                                                final MetricId metricId) {
+    final LoadingCache<Id, Meters> metersCache =
+        CacheBuilder.<Id, Meters>newBuilder()
+            .build(new CacheLoader<Id, Meters>() {
+              @Override
+              public Meters load(final Id id) throws Exception {
+                return Meters.create(registry, metricId.tagged("model", id.value()));
+              }
+            });
+
+    return new AutoValue_SemanticPredictorMetrics(metersCache);
   }
 
   @Override
   public PredictFnMetrics predictFnMetrics() {
-    return () -> SemanticPredictMetrics.create(predictDurationTimer().time(), predictRateCounter());
+    return id -> {
+      final Meters meters = metersCache().getUnchecked(id);
+      final Context time = meters.predictDurationTimer().time();
+      final Meter meter = meters.predictRateCounter();
+
+      return SemanticPredictMetrics.create(time, meter);
+    };
   }
 
   @Override
   public FeatureExtractorMetrics featureExtractorMetrics() {
-    return () -> SemanticVectorMetrics.create(extractDurationTimer().time(), extractRateCounter());
+    return id -> {
+      final Meters meters = metersCache().getUnchecked(id);
+      final Context time = meters.extractDurationTimer().time();
+      final Meter meter = meters.extractRateCounter();
+
+      return SemanticVectorMetrics.create(time, meter);
+    };
+  }
+
+  @AutoValue
+  abstract static class Meters {
```

Comment:
Should this be called `Metrics` instead of `Meters`?

Reply:
Alright, IMO `Metrics` makes more sense because this has both `Meter` and `Timer` objects but not a super strong opinion :)

Reply:
Alright, let's rename this.